ygot/render.go

```diff
@@ -1112,85 +1147,55 @@ func structJSON(s GoStruct, parentMod string, args jsonOutputConfig) (map[string
 			continue
 		}
 
-		for _, p := range mapPaths {
-			v, ok := value.(map[string]interface{})
-			switch p.Len() {
-			case 0:
-				if ok {
-					for mk, mv := range v {
-						k := mk
-						if appendModName {
-							// Append the module name for the 0th element if
-							// specified to do so. This is the module name of
-							// the root.
-							k = fmt.Sprintf("%s:%s", appmod, mk)
-						}
-						jsonout[k] = mv
-					}
-				} else {
-					errs.Add(fmt.Errorf("empty path specified for non-root entity"))
-					continue
-				}
-			case 1:
-				pelem, err := p.StringElemAt(0)
-				if err != nil {
-					errs.Add(err)
-					continue
+		if isFakeRoot {
+			if v, ok := value.(map[string]interface{}); ok {
+				for mk, mv := range v {
+					jsonout[mk] = mv
 				}
-				if appendModName {
-					pelem = fmt.Sprintf("%s:%s", appmod, pelem)
-				}
-				jsonout[pelem] = value
-			default:
-				var nilParent bool
-				parent := jsonout
-				for i := 0; i < p.Len()-1; i++ {
-					k, err := p.StringElemAt(i)
-					if err != nil {
-						errs.Add(err)
-						continue
-					}
+			} else {
+				errs.Add(fmt.Errorf("empty path specified for non-root entity"))
+			}
+			continue
+		}
 
-					switch {
-					case (i == 0 && appendModName && !p.isAbsolute):
-						// If the path is not absolute, then path compression has
-						// occurred - and therefore the elements must be in the
-						// same module. In this case, we append the module name
-						// to the first element in the list.
-						fallthrough
-					case i == 0 && appendModName && parentMod == "":
-						// For the first element, regardless of whether the path
-						// was absolute or not, we always must append the module
-						// name if there was no parent module, since this is an
-						// entity at the root.
-						k = fmt.Sprintf("%s:%s", appmod, k)
-						nilParent = true
-					}
-					if _, ok := parent[k]; !ok {
-						parent[k] = map[string]interface{}{}
-					}
-					parent = parent[k].(map[string]interface{})
-				}
-				k, err := p.LastStringElem()
+		if appmods != nil && len(mapPaths) != len(appmods) {
+			errs.Add(fmt.Errorf("%s: number of paths and modules in struct tag not the same: (paths: %v, modules: %v)", fType.Name, len(mapPaths), len(appmods)))
+			continue
+		}
+
+		for i, p := range mapPaths {
+			if appmods != nil && p.Len() != len(appmods[i]) {
+				errs.Add(fmt.Errorf("number of paths and modules elements not the same: (paths: %v, modules: %v)", p, appmods[i]))
+				continue
+			}
+
+			parent := jsonout
+			j := 0
+			for ; j != p.Len()-1; j++ {
```

Comment:
If `j` is being used outside of the loop context I'd call it something different. I think here something like `pathInMod`?

Reply:
I still like `j` better because I think it's easier to understand. The reason why it needs to be outside is because the last element is a special case (it needs to set the child JSON value), so I think its natural meaning as an index is not lost.

Reply:
SGTM :-)